modules/porous_flow/src/kernels/PorousFlowDarcyBase.C

```diff
@@ -120,27 +131,53 @@ PorousFlowDarcyBase::upwind(JacRes res_or_jac, unsigned int jvar)
       ((res_or_jac == JacRes::CALCULATE_JACOBIAN) ? _porousflow_dictator.porousFlowVariableNum(jvar)
                                                   : 0);
 
+  DenseMatrix<Number> & ke = _assembly.jacobianBlock(_var.number(), jvar);
+  if ((ke.n() == 0) &&
+      (res_or_jac == JacRes::CALCULATE_JACOBIAN)) // this removes a problem encountered in
+                                                  // the initial timestep when
+                                                  // use_displaced_mesh=true
+    return;
+
   /// The number of nodes in the element
   const unsigned int num_nodes = _test.size();
 
   /// Compute the residual and jacobian without the mobility terms. Even if we are computing the Jacobian
   /// we still need this in order to see which nodes are upwind and which are downwind.
-
-  std::vector<std::vector<Real>> component_re(num_nodes);
-  for (_i = 0; _i < num_nodes; ++_i)
+  for (unsigned ph = 0; ph < _num_phases; ++ph)
   {
-    component_re[_i].assign(_num_phases, 0.0);
+    _proto_flux[ph].assign(num_nodes, 0);
     for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      for (unsigned ph = 0; ph < _num_phases; ++ph)
-        component_re[_i][ph] += _JxW[_qp] * _coord[_qp] * darcyQp(ph);
+    {
+      for (_i = 0; _i < num_nodes; ++_i)
+        _proto_flux[ph][_i] += _JxW[_qp] * _coord[_qp] * darcyQp(ph);
+    }
   }
 
-  DenseMatrix<Number> & ke = _assembly.jacobianBlock(_var.number(), jvar);
-  if ((ke.n() == 0) &&
-      (res_or_jac == JacRes::CALCULATE_JACOBIAN)) // this removes a problem encountered in
-                                                  // the initial timestep when
-                                                  // use_displaced_mesh=true
-    return;
+  // for this element, record whether each node is "upwind" or "downwind" (or neither)
+  if (res_or_jac == JacRes::CALCULATE_RESIDUAL)
+  {
+    const unsigned elem = _current_elem->id();
+    if (_num_upwinds.find(elem) == _num_upwinds.end())
+    {
+      _num_upwinds[elem] = std::vector<std::vector<unsigned>>(_num_phases);
```

Comment:
New code paradigms to think about:
http://www.cplusplus.com/reference/unordered_map/unordered_map/emplace_hint/

This innocent line of code is actually doing a second lookup again since you just threw away the first lookup. It's not big deal, but I like to share fancy ways of doing things.